cmd/thanos/rule.go

```diff
@@ -327,26 +288,57 @@ func runRule(
 			}
 			alertQ.Push(res)
 		}
-
 		st := tsdb.Adapter(db, 0)
-		mgr = rules.NewManager(&rules.ManagerOptions{
-			Context:     ctx,
-			QueryFunc:   queryFn,
+
+		opts := rules.ManagerOptions{
 			NotifyFunc:  notify,
 			Logger:      log.With(logger, "component", "rules"),
 			Appendable:  st,
 			Registerer:  reg,
 			ExternalURL: nil,
 			TSDB:        st,
-		})
-		g.Add(func() error {
-			mgr.Run()
-			<-ctx.Done()
-			mgr.Stop()
-			return nil
-		}, func(error) {
-			cancel()
-		})
+		}
+
+		var refOpts *rules.ManagerOptions
+		{
+			ctx, cancel := context.WithCancel(context.Background())
```

Comment:
Maybe we should use wait groups here instead of context?

As this doesn't seem right to me:

```
g.Add(func() error {
mgr.Run()
<-ctx.Done()
mgr.Stop()
...
}
```

I would rather:

```
mgr = rules.NewManager(refOpts)
g.Add(func() error {
   return mgr.Run()
}, func(error) {
  mgr.Stop()
})
```